### ISSUE TITLE
Using old code in new release, java2html from 2006....

### ISSUE THREAD

**Issue by wltjr** [rzwitserloot/lombok#1437]
Please do not use old unmainted code in current projects. It makes it a nightmare for others to package. Even maven only provides binaries for de.java2html and not sources. The sources from upstream are a total mess.

Thus far having packaged hundreds of Java packages. Lombok is the absolute worst java package I have come across. It depends on a ridiculous amount of Eclipse, some 70+ packages. Not to mention bootstrap with lombok.patcher. Lombok seriously needs major work!!!

Now for a patch release, it seems 3 new dependencies are adding, java2html, freemarker, and another com.petebevin.markdown. Really should NOT be adding new deps on a patch release. Much less ancient ones like java2html. Seems there are some newer java2html on maven. Why de.java2html was used is beyond me. It has not been touched since 2006...

Lombok is making me hate Java. Please stop making a mess of this stuff... I am not surprised at all by the number of open bugs....

**Comment by wltjr** [comment]
Also  com.petebevin.markdown is the old namespace, per the release notes on 0.4
"Package name has been changed from com.petebevin.markdown to org.markdownj. The com.petebevin.markdown has been deprecated and will be removed in MarkdownJ 0.5."

The 0.4 package supports both namespaces, but for some reason Lombok went with com.petebevin vs org.markdownj. Which will break when 0.5 comes out. Likely better off to start with org.markdownj, and not require any changes when 0.5 comes out and com.petebevin is no more.

**Comment by victorwss** [comment]
I am not a developer of lombok, just a common user looking around, but... can you contextualize what you are doing?

BTW, lombok is not intended to be added as a runtime dependency to anything at all - it is intended to be used solely as a compile-time-only dependency used in annotation processing. Including lombok in the classpath of an end-user application is just wrong. Having lombok as a runtime dependency is as bad and wrong as having the javadoc tool as a runtime dependency.

I don't have an authoritative word here, but I think that one of the main reasons why lombok is not very modularized is because since it should not be a runtime dependency, it is more valuable as a single-jar download of a tool that just works (regardless of the fact that it is bloated) instead of something that has careful projected dependencies that could be upgraded independently.

Lombok is something special because it is a tool that somewhats abuses both ECJ and javac by invading their private internal APIs and messing with on-memory compiler ASTs that were expected to be to do its black magic. Also, it needs to integrate with Netbeans, Eclipse, IntelliJ, etc, and this adds a lot of bloat. So, I'm not surprised that it needs many things from Eclipse.

About markdownj specifically, I have no clue on that.

**Comment by wltjr** [comment]
I am packaging this stuff on Gentoo, [link](https://github.com/Obsidian-StudiosInc/os-xtoo/tree/master/dev-java/lombok).

**Comment by rspilker** [comment]
Java2html, freemarker and markdown are not in lombok.jar. They are in the build script to generate the website and the installer.

**Comment by rspilker** [comment]
"Lombok is making me hate Java. Please stop making a mess of this stuff... I am not surprised at all by the number of open bugs...." 

Be nice or go away

**Comment by wltjr** [comment]
Please code and package software in a manner that is nice to others. Any attitude is a result of frustration after spending hours to package lombok and related. Just to have to spend additional time on a patch release due to other changes. I [packaged hundreds of java packages](https://github.com/Obsidian-StudiosInc/os-xtoo/). Any lack of niceness is a result of not being nice or friendly to others in how the package is coded, and/or built.

I am not using the build script. I do a straight javac via a Gentoo eclass bash function. You can see some in the [ebuild for lombok](https://github.com/Obsidian-StudiosInc/os-xtoo/blob/master/dev-java/lombok/lombok-9999.ebuild). It finds actual usage in code, thus having to add packages as compile time dependencies. DEPEND is build time, and RDEPEND is runtime. CP_DEPEND is for classpath. Anything that is in both is needed both at compile an runtime. Build time stuff would only be added to DEPEND and not RDEPEND. Things like javacc, etc. Used to generate code but not build time dependencies.

I cannot recall the issue # but FYI that patch came from another issue I opened where I was nice, as was response etc. Thus I tend to conduct myself based on the situation. I was not happy to have to package additional things for lombok. I only had to package some 70+ eclipse packages as dependencies of lombok....

**Comment by wltjr** [comment]
FYI there are a few packages ~10 out of hundreds that really make me dislike java. Thus I do not make that statement lightly. Please do not take it personally as more feedback. It is not my mission, but it would be nice to see more Java packages be friendlier and built in a more consistent manner. Doing C coding and switching between say autotools, cmake, and meson. It is much more pleasant than dealing with say ant, maven, gradle. Mostly when it comes to things like dependencies etc. Java downloads other things tend to build. Thus many in java care less about packaging or others building the software. They just say use the jar and be happy. Status quo, but it can be improved.

Sorry for any harshness.

**Comment by rspilker** [comment]
I understand where you're coming from. No hard feelings, and thanks for all the work you put into this!

First of all, right now our main focus is on jdk9 support. So it will take some time to improve things.

I think you are refering to https://github.com/Obsidian-StudiosInc/os-xtoo/commit/dc0aa4cfc834ac598810a7db59ddcac5a8a0dbfc#diff-a58af25837303da01a5e9d698ac3ded5

In https://github.com/rzwitserloot/lombok/blob/master/buildScripts/ivy.xml we currently make no difference between:
- the dependencies we need to create the jar file
- the dependencies to create the website and documentation
- the dependencies we need to publish everything
I guess we could make those distinctions. Would that help you?

Also, we already use some stub files to compile against the different ecj/eclipse versions. We could probably do the same for more of the eclipse code, but that will take some time. We would probably write some more code to extract those from the existing jars. This is something we considered before, but never got to do it.

What is the purpose of your packaging? Will it only be used for compiling other code using javac? In that case, it would be possible to remove the whole eclipse part as well as the installer, which is a Swing GUI and uses some packaged windows dll files.

The reason we put it all in one repository and eventually in one jar is because for most users, that's exactly what they want: one thing that's small and can do it all.

**Comment by rspilker** [comment]
Oh, I forgot about the ecj-compiler. So not all "eclipse" code can be omitted, if you also want to support ecj.

**Comment by rspilker** [comment]
https://github.com/rzwitserloot/com.zwitserloot.stubber

**Comment by wltjr** [comment]
Yes that patch came from you referenced commit. We had a conversation there. I thought the patch came from you specifically. My main gripe was it seemed new dependencies were added in a patch release. Not sure why that was, I just was like "What?". I just packaged a bunch of stuff for lombok and now it needs more... If it was a minor or major version change, that would make a bit more sense. I would expect such changes. On patch release, x.x.N usually do not see such major changes. I surely would have noticed if it was needed to build as it ended up being with the latest release.

Thus I guess the solution to my issue/gripe is more in versioning. The distinctions is not so much necessary as I do not even pay attention to say pom.xml or build.gradles for depdencies. I care more about what the compiler needs. The rest comes later if its missing something for running or otherwise.

I guess a further gripe would be avoid adding new dependencies, code from other projects. That may not be practical always. If a project does have to be used, a current one. There are many ancient java packages still in use. Sources are hard to come by, and repos are in questionable states, etc. Many do not care as they just grab the ancient jar from maven or gradle. I would prefer if a package does add another as a new dependency. Its a current package, in good condition, actively developed, etc.

The stub file and other is a larger Java issue. Not blaming you all for the eclipse dependency chain. Not sure it can be avoided as this stuff even Lombok is a dependency of other stuff. So even with stubs I likely need the packages for actual usage in other stuff, Spring, etc. My goal is packaging other stuff, and the nature of Java. I end up packaging a lot of dependencies for any end goal package. Like Jenkins from source, needs Spring. Spring from source in its dependency chain needs Lombok, and many other things. I would blame Spring more than other stuff. it seems to depend on just about every Java package written indirectly :)

Not sure anything can be done to actually reduce lombok dependencies in my case. Not really so much the gripe as when they are added. Which comes back to versioning :) Something so small, can make such a difference. Also not adding old packages as new dependencies. Use new projects. 

The stubs and reducing deps bring other aspects. USE flags on gentoo control build options, including/excluding. There was talk of using stubs like pre-processor IF DEFINE in C/C++ to not have to have all dependencies present to build stuff. Though would need to be there for runtime. Some packages get around this by deleting files based on a USE flag, So say USE="-ecj" emerge lombok, as an example. That would require either stubs and/or deleting stuff that needed ecj, to avoid having ecj packaged and present on classpath, etc. That just ends up creating more work. Sometimes easier just to package the dependency, and based on other packages usage. Parts cannot be made optional it may need it all. I avoid the USE flag thing mostly for now and build packages as a whole, and all their dependencies...

I know I am a very abnormal Java user and packager. Gentoo is abnormal for most, much less building Java from source. I upset many doing things abnormally not even attitude or words. Some what why I do not worry about mine. If its just going to upset others with what I am doing, not saying. Run into various stuff as a result and the reception varies. Though the things I come across usually can benefit all. Java has some strange idioms. I am not trying to improve the world, but maybe a package at a time :)

**Comment by rzwitserloot** [comment]
If I can parse this feature request correctly, it is effectively not so much 'please update this dependency to something more recent', but more:

__Please do not change or add dependencies for minor version updates.__

Did I understand that correctly?

If so: This feature request is understood, and is __hereby denied__; as @rspilker tried to explain, absolutely nobody takes any notice of us updating certain kinds of deps, in any way or form, _except_ your unique use case. Handicapping ourselves like that serves the purposes of gentoo and not much else, and it seriously hampers our ability to produce updates.

If you can explain what you are trying to do, exactly, we can try to work out a good solution together. Please try not to use terms like 'CP_DEPEND'; you get really angry with the project lombok maintainers about not respecting the needs of your community, and yet you seem intent on treading on ours with disdain by using terms that aren't going to be familiar to java programmers that haven't tried to package for gentoo or another linux distro. Not everyone on earth is a gentoo maintainer; we have no idea what you are talking about, and your tone and attitude is making it very hard for me to work up the spirit to try to investigate.

Example of a thing I can try to help you with: "I am trying to write a script that turns the files in your github repo in a working lombok.jar, and this script cannot rely at all on any binary or internet download of any kind, it can only rely on binaries produced via a gentoo script (and such scripts themselves hold to the same rule: They can only pull in other gentoo scripts, and build from source, that's it)." I have no idea if that's what you're trying to do, I'm just guessing. I hope the idea of what I'm looking for is clear.

**Comment by wltjr** [comment]
To be clear, its two fold
1. Please do not add new dependencies to a patch release, consider doing a minor or major release. That way it is less of a surprise to others.
2. Please do not add old projects as new dependencies. When adding a new dependency, ideally check to see if it is an active project, etc. Old code and projects need that no one is furthering need not be used any more.

Regarding 1;
I am not sure why versioning would hamper ability to produce updates. This came as part of a version/tag, just a patch version. Which typically reflects no major or minor changes, just bugs fixed etc.

Regarding 2;
At times to lots of focus say on Java releases. Java 9 etc. While projects continue to use very old code that is not being updated to use new Java features. It starts to create a mess. Not sure what the point is for current projects to use old code. Long time dependency is one thing, but new is strange.

I was explaining what I was doing, and even some details thus mentioning CP_DEPEND. @rspilker had said the dependencies were not compile/build time, but other. I was saying that per my experience in building, and showing the technical details of how with some explanation of variables, not just opinion. The dependencies were needed to compile, build time dependencies. Sorry for any confusion of new terms. It is just a bash variable. I am not sure why what would be hard to understand. Shell variables are normal in any OS or env. Java relies on them, JAVA_HOME, etc.

In simple terms, this is packaging Lombok for Gentoo, as a dependency of other stuff, I believe Spring. Gentoo being a from source distribution, packages build from source. Not using pre-built binaries. Only in some rare cases where things bootstrap themselves. Ideally it would be great if less things bootstrapped. From a packaging and building from source perspective it is not friendly. If you recall past issue opened on lombok.patcher and lombok [circular dependencies](https://github.com/rzwitserloot/lombok.patcher/issues/2).

I had already spent considerable time to package a very large amount of dependencies for Lombok. More than I have seen for many packages, given it required a substantial portion of Eclipse. It was quite frustrating to on a patch release see new dependencies that required more time. Thus the tone.

**Comment by rzwitserloot** [comment]
The lombok circular dependency thing is something we'll reconsider at some point (it's on our list of things to do; just not particularly high on it).

I understand the source of your frustration. Doesn't change the fact that your 2 requests are both denied. I understand entirely that it's frustrating for you, but you're the only one who benefits with us granting these requests, and while you evidently can't see how it hampers development, I don't think it's worth the time to explain. It just does.

**Comment by wltjr** [comment]
Thanks for the consideration on the circular dependency issue. Though that I already worked around and is not an issue like the one raised and closed.

@rzwitserloot I some what expected the denial thus the tone and attitude. If you cared such things would not be done in the first place. Though at least the issue has been raised for the record. Though I rather a more honest we I rather not. If your concerned about tone.

If versioning effects your development. Not sure what to say, but versions are very cheap in git. This was a version release to begin with. If minor or major vs patch messes up your development.... That sounds very strange and likely not justifiable technically, beyond personal preference.

As for continuing to use old code. Clearly you do not care about the Java eco system. Nor weak links for Lombok. java2html was never even updated for Java 1.5. It is that old. Added to Lombok in the same year Java 9 is to be released. Most projects are looking forward. if Lombok wants to live in the past, using projects of yester year that is surely logical. But almost no one supports 1.5, much less 1.4... Even vendors have moved on. Oracle will not support 1.5, nor IBM or Redhat.

When coming across such, almost knowing it will fall on deaf ears not to be changed. Tone is more justified in my opinion. This just creates a mess in the Java eco system.
